File: src/main/java/io/github/oxnz/Ingrid/tx/TxService.java
package io.github.oxnz.Ingrid.tx;

import org.apache.http.client.methods.HttpPost;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import javax.persistence.NoResultException;
import java.io.IOException;
import java.util.Set;
import java.util.concurrent.ExecutionException;

@Service
public class TxService implements AutoCloseable {

    final Logger log = LoggerFactory.getLogger(getClass());

    private final TxRecordRepo txRecordRepo;
    private final TxResultRepo txResultRepo;
    private final Dispatcher dispatcher;
    private final HttpExecutionService httpExecutionService;

    public TxService(TxRecordRepo txRecordRepo, TxResultRepo txResultRepo, Dispatcher dispatcher, HttpExecutionService httpExecutionService) {
        this.txRecordRepo = txRecordRepo;
        this.txResultRepo = txResultRepo;
        this.dispatcher = dispatcher;
        this.httpExecutionService = httpExecutionService;
    }

    public void process(long id) throws TxException {
        try {
            TxRecord record = txRecordRepo.findById(id).orElseThrow(NoResultException::new);
            log.debug("record: {}", record);
            // TODO: cxService
            // cxRecord cxRecord = cxService.complete(record.getId());
            // record.data = cxRecord.toString();
            // txRecordRepo.save(record);
            Set<DestSpec> destSpecs = dispatcher.dispatch(record);
            destSpecs.forEach(destSpec -> post(record, destSpec));
        } catch (Exception e) {
            throw new TxException(e);
        }
    }

    private void post(TxRecord record, DestSpec destSpec) {
        HttpExecutionResult httpExecResult = doPost(record, destSpec);
        TxResult txResult = new TxResult(record, httpExecResult.succeeded(), httpExecResult.message());
        txResultRepo.save(txResult);
        log.debug("result: {}", txResult);
    }

    private HttpExecutionResult doPost(TxRecord record, DestSpec destSpec) {
        HttpPost request = destSpec.requestBuilder().buildRequest(record, destSpec);
        try {
            return httpExecutionService.execute(request, null, destSpec.responseHandler());
        } catch (InterruptedException | ExecutionException e) {
            log.error("post", e);
            return new HttpExecutionResult(false, e.getMessage());
        }
    }

    @Override
    public void close() throws IOException {
        httpExecutionService.close();
    }
}

File: src/main/scala/io/github/oxnz/Ingrid/tx/TxService.scala
package io.github.oxnz.Ingrid.tx

import java.io.IOException

import io.github.oxnz.Ingrid.cx.{CxRequest, CxResponse, CxService}
import javax.persistence.NoResultException
import org.slf4j.{Logger, LoggerFactory}
import org.springframework.stereotype.Service

import scala.util.{Failure, Success, Try}

@Service class TxService(val txRecordRepo: TxRecordRepo, val txResultRepo: TxResultRepo, val cxService: CxService, val dispatcher: TxDispatcher, val txHttpExecutor: TxHttpExecutor) extends AutoCloseable {
  final private[tx] val log: Logger = LoggerFactory.getLogger(getClass)

  @throws[TxException]
  def process(id: Long): Unit = {
    val record: TxRecord = txRecordRepo.findById(id).orElseThrow(() => new NoResultException("record not found: " + id))
    log.debug("record: {}", record)
    try {
      record.status = TxStatus.CONSUMED
      val cxRequest: CxRequest = new CxRequest(record.cat, record.ref)
      val cxResponse: CxResponse = cxService.process(cxRequest)
      if (!cxResponse.succ) throw new IllegalStateException("cxService failed")
      record.status = TxStatus.COMPLETED
      record.data = cxResponse.data
      txRecordRepo.save(record)
      val destSpecs: Set[TxDestSpec] = dispatcher.dispatch(record)
      log.debug("dispatch: {} => {}", record, destSpecs)
      destSpecs.foreach(process(record, _))
      record.status = TxStatus.SENT
    } catch {
      case e: Exception =>
        record.status = TxStatus.FAILED
        throw new TxException(e)
    } finally txRecordRepo.save(record)
  }

  private def process(record: TxRecord, destSpec: TxDestSpec): Unit = {
    val httpExecResult: TxHttpExecResult = transfer(record, destSpec) match {
      case Success(result) => result
      case Failure(exception) =>
        log.error("transfer", exception)
        new TxHttpExecResult(false, exception.getMessage)
    }
    val txResult: TxResult = new TxResult(record, httpExecResult.succ, httpExecResult.msg)
    txResultRepo.save(txResult)
    log.debug("result: {}", txResult)
  }

  private def transfer(record: TxRecord, destSpec: TxDestSpec): Try[TxHttpExecResult] = Try {
    val request = destSpec.requestBuilder.buildRequest(record, destSpec)
    txHttpExecutor.execute(request, destSpec.httpContext, destSpec.responseHandler)
  }

  @throws[IOException]
  override def close(): Unit = {
    txHttpExecutor.close()
  }
}
